### PR TITLE
[haproxy] unix socket stats urls

### PR DESCRIPTION
**Pull request by sj26**
### What does this PR do?

Support unix sockets for haproxy stats collection.

### Motivation

Running a separate stats socket with tight permissions instead of over a tcp port provides greater security.

### Testing Guidelines

Sorry, I have no idea how to get the tests actually working in a local environment. It has been a world of pain. I'm hoping travis does a better job at running them and providing some useful feedback.

I'm on macOS 10.12, using homebrew, ruby 2.3.1, python 2.7.12. `rake` aborts on rubocop errors. So does `rake ci:run`. `rake ci:run[haproxy]` tries to compile haproxy which can't find openssl, even if I supply correct CFLAGS and LDFLAGS to my homebrew'd openssl install as env. Subsequent `rake ci:run[haproxy]` runs don't try to recompile haproxy, they fail with `Errno::ENOENT No such file or directory - /Users/sj26/Projects/dd-agent/embedded/haproxy_1.5.11/haproxy`. And cleanup fails because it can't find an haproxy pid.

I haven't attempted to figure out mocks yet.

### Additional Notes

I realise this introduces a new dependency which requires updating your omnibus, but I'd like to get the code working first before jumping on that dragon.

### CODE REVIEW

**Comment by masci** [comment]
Hi @sj26 and thanks for your code, we now that running the tests locally is not as easy as it should be, sorry for that!

Aside from the details that are making the CI fail, I'm a little bit concerned about the overall approach - if I understood correctly, by `monkey_patch`ing requests you're changing the behaviour of the check entirely, switching from TCP to unix sockets.

I agree moving to unix socket would improve security but we cannot break the compatibility with existing configurations. I'd make the check compatible with both approaches, defaulting to the old behaviour and making available the new one through the config file.

What do you think?

**Comment by sj26** [comment]
The monkey patch shouldn't prevent requests from functioning as normal, or change any existing integrations, it just adds an extra adapter for the "http+unix" scheme. :-)

File: ci/haproxy.rb [hunk 1]
```diff
@@ -44,8 +44,13 @@ def haproxy_rootdir
...
     task before_script: ['ci:common:before_script'] do
       %w(haproxy haproxy-open).each do |name|
+        # Older haproxy doesn't support ENV interpolation
+        config = File.read("#{ENV['TRAVIS_BUILD_DIR']}/ci/resources/haproxy/#{name}.cfg")
+        config.gsub!("$VOLATILE_DIR", ENV["VOLATILE_DIR"])
+        File.write("#{ENV["VOLATILE_DIR"]}/#{name}.cfg", config)
+
         pid = spawn("#{haproxy_rootdir}/haproxy", '-d', '-f',
-                    "#{ENV['TRAVIS_BUILD_DIR']}/ci/resources/haproxy/#{name}.cfg",
+                    "#{ENV["VOLATILE_DIR"]}/#{name}.cfg",
```
Comment (masci) [inline ci/haproxy.rb]: change these to single quotes to make rubocop stop complaining

**Comment by masci** [comment]
Nice, that was my impression looking at the CI but I discovered the existence of `requests_unixsocket` only this morning 😄

File: checks.d/haproxy.py [hunk 2]
```diff
@@ -147,19 +158,38 @@ def check(self, instance):
...
+
+        response = ""
+        output = sock.recv(8192)
+        while output:
+            response += output.decode("ASCII")
+            output = sock.recv(8192)
 
-        r = requests.get(url, auth=auth, headers=headers(self.agentConfig), verify=verify, timeout=self.default_integration_http_timeout)
-        r.raise_for_status()
+        sock.close()
```
Comment (sj26) [inline checks.d/haproxy.py]: Most of this was borrowed from https://github.com/neurogeek/haproxyctl/blob/master/haproxy/conn.py#L51-L83

**Comment by sj26** [comment]
@masci sorry for the delays.

Turns out I was completely wrong and the admin/stats sockets don't run HTTP at all but uses their own line-based protocol. I've updated the implementation to accept `unix:///var/run/haproxy.sock` style values and talk unix sockets directly instead of using http in those cases. It actually makes things easier, and we don't need another egg.

Take another look?

File: checks.d/haproxy.py [hunk 3]
```diff
@@ -147,19 +158,38 @@ def check(self, instance):
...
+        ''' Hit a given stats socket and return the stats lines '''
+
+        self.log.debug("Fetching haproxy stats from socket: %s" % socket_path)
+
+        sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
+        sock.connect(socket_path)
+        sock.send("show stat\r\n")
+
+        response = ""
+        output = sock.recv(8192)
```
Comment (masci) [inline checks.d/haproxy.py]: Can you define this number as a constant on the top of the module? Like:
`BUFSIZE=8192`

File: conf.d/haproxy.yaml.example [hunk 4]
```diff
@@ -5,6 +5,10 @@ instances:
     # username: username
     # password: password
     #
+    # or, with a unix stats or admin socket:
+    #
+    # url: unix:///var/run/haproxy.sock
```
Comment (masci) [inline conf.d/haproxy.yaml.example]: Nitpick: could you add a leading `-` to the example so it's clear that should be another instance and not an attribute of the first one?
```
 instances:
   - url: http://localhost/admin?stats
      # username: username
      # password: password
   # or, with a unix stats or admin socket:
   # - url: unix:///var/run/haproxy.sock
```

File: checks.d/haproxy.py [hunk 5]
```diff
@@ -6,7 +6,12 @@
 from collections import defaultdict
 import copy
 import re
+import socket
 import time
+try:
+    import urlparse
+except LoadError:
```
Comment (sj26) [inline checks.d/haproxy.py]: :+1:

File: checks.d/haproxy.py (hunk 3 above)
Comment (sj26) [inline checks.d/haproxy.py]: :+1:

File: conf.d/haproxy.yaml.example (hunk 4 above)
Comment (sj26) [inline conf.d/haproxy.yaml.example]: Hmm, that looks odd to me, but updated.

Comment (masci) [inline conf.d/haproxy.yaml.example]: Yeah at first I thought it'd be clearer but doesn't seem so...
Feel free to adjust it as you think it's better from the user perspective, or leave it as it is and we'll fix that later.

**Comment by sj26** [comment]
Thanks!

**Pull request merged**